src/Experimental/Traversals/dfs.jl

```diff
@@ -11,36 +11,41 @@ function traverse_graph!(
     state::AbstractTraversalState,
     neighborfn::Function=outneighbors
     ) where U<:Integer
-    
+
     n = nv(g)
     visited = falses(n)
+    parents = zeros(U, n)
     S = Vector{U}()
-    sizehint!(S, length(ss))
     @inbounds for s in ss
         us = U(s)
         visited[us] = true
         push!(S, us)
         initfn!(state, us) || return false
     end
+
     while !isempty(S)
         v = S[end]
-        previsitfn!(state, v) || return false
-        new_neighbor_found = false
-        @inbounds for i in neighborfn(g, v)
-            visitfn!(state, v, i) || return false
-            if !visited[i]  # find the first unvisited neighbor
-                newvisitfn!(state, v, i) || return false
-                new_neighbor_found = true
+        if parents[v] != zero(U)
+            newvisitfn!(state, visited, parents[v], v) || return false
+        end
+        previsitfn!(state, visited, v) || return false
+
+        @inbounds for i in neighbors(g, v)
+            visitfn!(state, visited, v, i) || return false
+            if !visited[i]
                 visited[i] = true
+                parents[i] = v
                 push!(S, i)
-                break
             end
         end
-        postvisitfn!(state, v) || return false
-        if !new_neighbor_found  # no more new neighbors. Let's go to the next source vertex.
-            postlevelfn!(state) || return false
+
+        while (!isempty(S) && S[end] == v) 
+            postvisitfn!(state, visited, v) || return false
             pop!(S)
+            v = parents[v]
         end
+
+        # postlevelfn!(state, visited, v)
```

Comment:
My concern is that postlevelfn! may be run before postvisitfn!, because we only run postvisitfn! after all descendants have been explored, whereas postlevelfn! here runs after all children nodes have been scanned without further exploring their descendants. Should this be the expected behavior?